src/Storages/ExternalStream/Kafka/ConsumerPool.h

```diff
@@ -52,16 +54,25 @@ class ConsumerPool : public IConsumerPool, private PoolBase<Consumer>
     {
     }
 
+    void shutdown() override
+    {
+        if (stopped.test_and_set())
+            return;
+
+        LOG_INFO(log, "Shutting down consumer pool, waiting for all consumers to be freed");
+        waitForNoMoreInUse();
+        LOG_INFO(log, "All consumers are freed");
+    }
+
     Entry get(Int64 max_wait_ms) override
```

Comment:
Look like the interface is not well designed. It shall return `std::optional<Entry>`

Reply:
Because `Entry` already acts like a `std::optional` (  `Entry` has a `bool is_null()` function ), that's why I didn't use optional here. Otherwise, the code will need to do something like `if (result && !result->is_null())`. Hopefully this makes sense.

Reply:
it is a smell design in the base, let's skip fixing it for now.